src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -1025,6 +1065,221 @@ private static string GetFirstLineSubString(string stringToComplete, out bool ha
+                return true;
+            }
+
+            // Then try to complete for an enum type
+            if (!typeConstraint.IsEnum)
+            {
+                return false;
+            }
+            else
+            {
+                completions = GetResultForEnum(typeConstraint, completionContext);
+                return true;
+            }
+        }
+
+        private static List<CompletionResult> GetResultForSet(
+            Type type,
+            IList<string> validValues,
+            CompletionContext completionContext)
+        {
+            var allValues = new List<string>();
+            foreach (string value in validValues)
+            {
+                if (type == typeof(string) || type.IsEnum)
+                {
+                    allValues.Add(GetQuotedString(value, completionContext));
+                }
+                else
+                {
+                    allValues.Add(value);
+                }
+            }
+
+            return GetMatchedResults(allValues, completionContext);
+        }
+
+        private static List<CompletionResult> GetMatchedResults(
+            List<string> allValues,
+            CompletionContext completionContext)
+        {
+            var stringToComplete = string.Empty;
+            if (completionContext.TokenAtCursor != null)
+            {
+                stringToComplete = completionContext.TokenAtCursor.Text;
+            }
+
+            string matchString = stringToComplete + "*";
+            var wildcardPattern = WildcardPattern.Get(matchString, WildcardOptions.IgnoreCase | WildcardOptions.CultureInvariant);
+
+            IEnumerable<string> matchedResults = allValues.Where(r => wildcardPattern.IsMatch(r));
+            if (matchedResults == null || !matchedResults.Any())
+            {
+                matchedResults = allValues;
+            }
+
+            var result = new List<CompletionResult>();
+            foreach (var match in matchedResults)
+            {
+                result.Add(new CompletionResult(match));
+            }
+
+            return result;
+        }
+
+        private static string GetQuotedString(
+            string value,
+            CompletionContext completionContext)
+        {
+            var stringToComplete = string.Empty;
+            if (completionContext.TokenAtCursor != null)
+            {
+                stringToComplete = completionContext.TokenAtCursor.Text;
+            }
+
+            var quote = stringToComplete.StartsWith('"') ? "\"" : "'";
+            return quote + value + quote;
+        }
+
+        private static List<CompletionResult> GetResultForEnum(
+            Type type,
+            CompletionContext completionContext)
+        {
+            var allNames = new List<string>();
+            foreach (var name in Enum.GetNames(type))
+            {
+                allNames.Add(GetQuotedString(name, completionContext));
+            }
+
+            allNames.Sort();
```

Comment:
Maybe we shouldn't sort here. The returned names are very likely in the order of how members are declared (not guaranteed though, see https://docs.microsoft.com/en-us/dotnet/api/system.enum.getnames?view=netframework-4.8#remarks).
I think it's better to complete in the declaration order.

Reply:
The feedback is that since the completer is returning strings, it's better to sort them.  The enum value is not of interest to most users.

Reply:
It turns out argument completion for `-ErrorAction` sorts the enum names too. So yes, let's do the sorting here.